osf/utils/machines.py

```diff
@@ -344,118 +346,127 @@ def get_context(self):
-        self.machineable.comment = self.action.comment
-        self.machineable.save()
+    SanctionStateMachine also provides some extra functionality to write
+    RegistrationActions on events moving in to or out of Moderated machine states
+    as well as to convert MachineErrors (which arise on unsupported state changes
+    requests) into HTTPErrors to report back to users who try to initiate such errors.
+    '''
 
-    def accept_draft_registration(self, ev):
-        approval = self.machineable.approval
-        approval.approve(self.action.creator)
-        approval.save()
+    SANCTION_TYPE = SanctionTypes.UNDEFINED
+    MACHINE_STATE_FIELD_NAME = ''
 
-    def reject_draft_registration(self, ev):
-        self.machineable.meta = {}
-        self.machineable.save()
+    def __init__(self):
 
-    def submit_draft_registration(self, ev):
-        embargo_date = ev.kwargs.get('embargo_date', None)
-
-        if embargo_date:
-            submission = 'embargo'
-            assert embargo_date, 'must include embargo date'
-        else:
-            submission = 'immediate'
-
-        self.machineable.submit_for_review(
-            self.action.creator,
-            {
-                'registration_choice': submission,
-                'embargo_end_date': embargo_date
-            },
-            save=True
+        super().__init__(
+            states=SanctionStates,
+            transitions=SANCTION_TRANSITIONS,
+            initial=SanctionStates.from_db_name(getattr(self, self.MACHINE_STATE_FIELD_NAME)),
+            model_attribute='approval_stage',
+            prepare_event='_initialize_transition',
+            after_state_change='_save_transition',
+            send_event=True,
+            queued=True,
         )
 
-    def withdrawal_request_fails(self, ev):
-        user = ev.kwargs.get('user')
-        self.machineable.registered_node.retraction._on_reject(user)
-        self.machineable.registered_node.retraction.state = Retraction.REJECTED
-
-    def force_withdrawal(self, ev):
-        pass
-
-    def notify_embargo_termination(self, ev):
-        pass
-
-    def notify_embargo(self, ev):
-        pass
-
-    def notify_withdraw(self, ev):
-        pass
-
-    def notify_accept_reject(self, ev):
-        pass
-
-    def notify_edit_comment(self, ev):
-        pass
-
-    def notify_resubmit(self, ev):
-        pass
-
-    def notify_submit(self, ev):
-        pass
+    @property
+    def target_registration(self):
+        raise NotImplementedError(
+            'SanctionStateMachine subclasses must define a target_registration property'
+        )
 
-    def notify_withdraw_request_submitted(self, ev):
-        pass
+    @property
+    def approval_stage(self):
+        raise NotImplementedError(
+            'SanctionStateMachine subclasses must define an approval_stage property with a setter.'
+        )
 
-    def notify_withdraw_request_denied(self, ev):
-        pass
+    def _process(self, *args, **kwargs):
```

Comment:
The existing _run_transition function in Machinable wraps the state machine call in `with transaction.atomic` context manager. I *think* we could successfully do that here, too, if it's necessary.

Reply:
Worth noting transactions are already atomic within a single request: https://docs.djangoproject.com/en/3.1/topics/db/transactions/#tying-transactions-to-http-requests